### ISSUE TITLE
ERROR: Parent node is busy setting up children, `add_child()` failed. Consider using `add_child.call_deferred(child)` instead.

### ISSUE THREAD

**Issue by Firezac** [godotengine/godot#77095]
Hello!
I'm making a game at Godot that plays with first-person Parkour mechanics and somehow, shortly after I messed with the resolution settings of my project, this bug started showing up in my project both in the console and in Debugger.
The biggest problem with this bug is that it doesn't allow me to open my Project Settings. Every time I try, the engine crashes.
I first reported it on the Godot Facebook group, but nobody found the problem. Some users pointed out the possibility of some plugin causing this problem, but the only plugin I'm using is a Kanban plugin that I've had since the beginning of the project it didn't give problems before and the problem doesn't go away after I delete it.
I haven't found other people with the same problem, only this case (https://github.com/godotengine/godot/issues/73765) in which the same bug happens in a different way and by triggering an option that is not activated in my project.
I am convinced that this is some structural bug in the engine because the more I spindle my project and all my scripts, the less I find reasons for it to happen, considering that it appears every time the game is started.
Anyway, I'm sending here the code from my Main Scene, my project.godot and my global script, which are nothing big, just basic scripts to set the game menu (after all, the bug appears as soon as the engine starts, before any scene is started).

project.godot:
```
; Engine configuration file.
; It's best edited using the editor UI and not directly,
; since the parameters that go here are not all obvious.
;
; Format:
;   [section] ; section goes between []
;   param=value ; assign values to parameters

config_version=5

[application]

config/name="FPSTest"
run/main_scene="res://main_game_root.tscn"
config/features=PackedStringArray("4.0", "Forward Plus")
config/icon="res://icon.svg"

[autoload]

Global="*res://Global.gd"

[display]

window/size/resizable=false
window/stretch/mode="canvas_items"

[editor_plugins]

enabled=PackedStringArray("res://addons/kanban_tasks/plugin.cfg")

[input]

ui_accept={
"deadzone": 0.5,
"events": [Object(InputEventKey,"resource_local_to_scene":false,"resource_name":"","device":0,"window_id":0,"alt_pressed":false,"shift_pressed":false,"ctrl_pressed":false,"meta_pressed":false,"pressed":false,"keycode":32,"physical_keycode":0,"key_label":0,"unicode":32,"echo":false,"script":null)
]
}
ui_select={
"deadzone": 0.5,
"events": [Object(InputEventKey,"resource_local_to_scene":false,"resource_name":"","device":-1,"window_id":0,"alt_pressed":false,"shift_pressed":false,"ctrl_pressed":false,"meta_pressed":false,"pressed":false,"keycode":0,"physical_keycode":4194325,"key_label":0,"unicode":0,"echo":false,"script":null)
]
}
ui_focus_next={
"deadzone": 0.5,
"events": [Object(InputEventKey,"resource_local_to_scene":false,"resource_name":"","device":-1,"window_id":0,"alt_pressed":false,"shift_pressed":false,"ctrl_pressed":false,"meta_pressed":false,"pressed":false,"keycode":0,"physical_keycode":4194326,"key_label":0,"unicode":0,"echo":false,"script":null)
]
}
ui_left={
"deadzone": 0.5,
"events": [Object(InputEventKey,"resource_local_to_scene":false,"resource_name":"","device":0,"window_id":0,"alt_pressed":false,"shift_pressed":false,"ctrl_pressed":false,"meta_pressed":false,"pressed":false,"keycode":4194319,"physical_keycode":0,"key_label":0,"unicode":4194319,"echo":false,"script":null)
, Object(InputEventJoypadButton,"resource_local_to_scene":false,"resource_name":"","device":0,"button_index":13,"pressure":0.0,"pressed":false,"script":null)
, Object(InputEventKey,"resource_local_to_scene":false,"resource_name":"","device":-1,"window_id":0,"alt_pressed":false,"shift_pressed":false,"ctrl_pressed":false,"meta_pressed":false,"pressed":false,"keycode":0,"physical_keycode":65,"key_label":0,"unicode":97,"echo":false,"script":null)
]
}
ui_right={
"deadzone": 0.5,
"events": [Object(InputEventKey,"resource_local_to_scene":false,"resource_name":"","device":0,"window_id":0,"alt_pressed":false,"shift_pressed":false,"ctrl_pressed":false,"meta_pressed":false,"pressed":false,"keycode":4194321,"physical_keycode":0,"key_label":0,"unicode":4194321,"echo":false,"script":null)
, Object(InputEventJoypadButton,"resource_local_to_scene":false,"resource_name":"","device":0,"button_index":14,"pressure":0.0,"pressed":false,"script":null)
, Object(InputEventKey,"resource_local_to_scene":false,"resource_name":"","device":-1,"window_id":0,"alt_pressed":false,"shift_pressed":false,"ctrl_pressed":false,"meta_pressed":false,"pressed":false,"keycode":0,"physical_keycode":68,"key_label":0,"unicode":100,"echo":false,"script":null)
]
}
ui_up={
"deadzone": 0.5,
"events": [Object(InputEventKey,"resource_local_to_scene":false,"resource_name":"","device":0,"window_id":0,"alt_pressed":false,"shift_pressed":false,"ctrl_pressed":false,"meta_pressed":false,"pressed":false,"keycode":4194320,"physical_keycode":0,"key_label":0,"unicode":4194320,"echo":false,"script":null)
, Object(InputEventJoypadButton,"resource_local_to_scene":false,"resource_name":"","device":0,"button_index":11,"pressure":0.0,"pressed":false,"script":null)
, Object(InputEventKey,"resource_local_to_scene":false,"resource_name":"","device":-1,"window_id":0,"alt_pressed":false,"shift_pressed":false,"ctrl_pressed":false,"meta_pressed":false,"pressed":false,"keycode":0,"physical_keycode":87,"key_label":0,"unicode":119,"echo":false,"script":null)
]
}
ui_down={
"deadzone": 0.5,
"events": [Object(InputEventKey,"resource_local_to_scene":false,"resource_name":"","device":0,"window_id":0,"alt_pressed":false,"shift_pressed":false,"ctrl_pressed":false,"meta_pressed":false,"pressed":false,"keycode":4194322,"physical_keycode":0,"key_label":0,"unicode":4194322,"echo":false,"script":null)
, Object(InputEventJoypadButton,"resource_local_to_scene":false,"resource_name":"","device":0,"button_index":12,"pressure":0.0,"pressed":false,"script":null)
, Object(InputEventKey,"resource_local_to_scene":false,"resource_name":"","device":-1,"window_id":0,"alt_pressed":false,"shift_pressed":false,"ctrl_pressed":false,"meta_pressed":false,"pressed":false,"keycode":0,"physical_keycode":83,"key_label":0,"unicode":115,"echo":false,"script":null)
]
}
Dash={
"deadzone": 0.5,
"events": [Object(InputEventKey,"resource_local_to_scene":false,"resource_name":"","device":-1,"window_id":0,"alt_pressed":false,"shift_pressed":false,"ctrl_pressed":false,"meta_pressed":false,"pressed":false,"keycode":0,"physical_keycode":67,"key_label":0,"unicode":99,"echo":false,"script":null)
]
}
Jump={
"deadzone": 0.5,
"events": [Object(InputEventKey,"resource_local_to_scene":false,"resource_name":"","device":-1,"window_id":0,"alt_pressed":false,"shift_pressed":false,"ctrl_pressed":false,"meta_pressed":false,"pressed":false,"keycode":0,"physical_keycode":32,"key_label":0,"unicode":32,"echo":false,"script":null)
]
}
Pause={
"deadzone": 0.5,
"events": [Object(InputEventKey,"resource_local_to_scene":false,"resource_name":"","device":-1,"window_id":0,"alt_pressed":false,"shift_pressed":false,"ctrl_pressed":false,"meta_pressed":false,"pressed":false,"keycode":0,"physical_keycode":4194305,"key_label":0,"unicode":0,"echo":false,"script":null)
]
}

[physics]

common/physics_ticks_per_second=250


```
Main scene Script:
```
extends Control

@export var lastScene: String
@export var actualScene: String
@export var nextScene: String
var instanceScene

@export var loadingScene = load("res://HUD/load_screen.tscn")
var instancedLoadingScene = loadingScene.instantiate()

var gameVis
var configVis
var menuVis
var loadVis

func _ready():
	pass

func _process(_delta):
	_menuScreen()
	_pauseScreen()
	_configScreen()
	
	gameVis = Global.gameVis
	configVis = Global.configVis
	menuVis = Global.menuVis
	loadVis = Global.loadVis
	
	nextScene = Global.nextSceneToLoad
	
	if Global.loadIsReady == true:
		if $SubViewportContainer/SubViewport/GameLayer.get_child_count() == 0:
			instanceScene = load(nextScene).instantiate()
			$SubViewportContainer/SubViewport/GameLayer.add_child(instanceScene)
			Global.loadIsReady = false
			Global.gameVis = true
			Global.configVis = false
			Global.menuVis = false
			Global.loadVis = false
			$SubViewportContainer/SubViewport/Background.visible = false
		else:
			Global.loadIsReady = false
	
	if Global.isItLoading == true:
		if $SubViewportContainer/SubViewport/LoadingLayer.get_child_count() == 0:
			$SubViewportContainer/SubViewport/LoadingLayer.add_child(instancedLoadingScene)
			Global.isItLoading = false
		else:
			Global.isItLoading = false

	$SubViewportContainer/SubViewport/GameLayer.visible = gameVis
	$SubViewportContainer/SubViewport/ConfigMenuLayer.visible = configVis
	$SubViewportContainer/SubViewport/MenuLayer.visible = menuVis
	$SubViewportContainer/SubViewport/LoadingLayer.visible = loadVis

func _menuScreen():
	pass

func _pauseScreen():
	pass

func _configScreen():
	pass

func _loadingScreen():
	pass

```	
Global Script:
```
extends Node

var gameVis = false
var configVis = false
var menuVis = true
var loadVis = false

var nextSceneToLoad = "res://HUD/load_screen.tscn"
var loadIsReady = false
var isItLoading = false

var pauseState = false

```
Plus, The error after I play the game:
```
Parent node is busy setting up children, `add_child()` failed. Consider using `add_child.call_deferred(child)` instead.
scene/main/window.cpp:1459 - Condition "!is_inside_tree()" is true.
Parent node is busy setting up children, `add_child()` failed. Consider using `add_child.call_deferred(child)` instead.
Parent node is busy setting up children, `add_child()` failed. Consider using `add_child.call_deferred(child)` instead.
```

**Comment by Firezac** [comment]
A new update about this bug: IT IS HAPPENING EVEN IN NEW UNTOUCHED PROJECTS!!!!
I was so frustrated trying to delete specific things in the programming to correct this bug, but nothing was working. Then, I just started a new project empty and just created a node. Nothing more than that, and the bug happened! I don't know if it is some problem with my system specifically but I really don't know what I can do.

**Comment by RichardR01** [comment]
Since update to 4.03 stable, I am getting it as well. It is happening now; where previously, it never happened. I am calling 2 functions that instantiate objects preloaded in another script that is autoloaded.

**Comment by Firezac** [comment]
I think I discovered what is happening, it is probably a cache conflict.
Like I said in my last post, this error was blocking me to access my project settings even when I just started a new project. I was curious about it because, in the (https://github.com/godotengine/godot/issues/73765) case, the developer was doing completely different things and ended up finding the same error.
Well, I was working on another aspect of my game and I had an error that recently stopped the engine and restarted it (I couldn't take the error because was too fast), but when I come back, I could back to access my game settings but now, my export function was not working.
So, I think "It's the same project as before, with nothing changed. What could have changed this to "move the error from one point to another"? Then I realize a thing: The cache! 
Every time the cache changes, no matter what I do. The project is sometimes still the same, but these errors change place and the only thing that the GitHub Desktop accuses of change is the cache.
So, I made a try and deleted the .godot folder in my project before starting my game, and god, it works!!!!!! It's not a definitive solution but for now, I think it can save many projects from everyone who is working at Godot 4 right now and find this error.
I am not a programming engineer so I am not able to go further at this point, but I really help this discovery help the devs for when Godot 4.0.4 releases. Maybe creating a system that restarts the old cache every time the project starts? Just a suggestion.

**Comment by KoBeWi** [comment]
Can you attach some minimal reproduction project?

**Comment by Firezac** [comment]
Actually, yes, can send you the archives. Not just that but I discovered that if I delete all the cache from the engine folder, not just the project folder, the bug will be correct at all. I will send here 2 zip files, one is my old cache from my engine. I believe if you put it in your cache folder, the bug will come back.
The other one is my .godot folder from my project, so you can try it in one empty project and see this bug. I can also send you an older version of my game with the bug if necessary, but I would be more comfortable sending it to you privately because, it's not a biiiiig and complex project, but I am working on it and don't want to put it online yet.
Considering it is too big to upload here, I am sending those files at the drive in this link (it's late right now, so I am letting upload it. If you don't two zip files in the folder, wait a bit and they probably will appear): https://drive.google.com/drive/folders/1cwPocqiefEqm1nkgX1AkuQXgptmtsBds?usp=drive_link

**Comment by KoBeWi** [comment]
> The other one is my .godot folder from my project, so you can try it in one empty project and see this bug.

Well, I don't see this bug. Although I didn't download the 2GB file .-.
It doesn't sound like an error that would be triggered by project cache. By minimal reproduction project I meant some scene/script that exhibits the problem.

**Comment by Firezac** [comment]
Well, I don't know what to think about it then. I'll try to list the events in order so we can try to find a pattern:

-This bug just appeared on my Godot 4.0.3 engine without any reason and stayed there even I start a new project. I turn on the engine and there it is, crashing the program when I try to access the editor or the export window.
-When I deleted the .godot folder from my project (the one that is in the link that I send you), the bud stops to block my settings but still showed on the debugger.
-When I delete the cache folder from my own Godot, the bug stops in any project and shows in the debugger.
-(a today update) After some time, the bug reappeared in the debugger without locking the settings windows, but redoing the process "corrected" it again.

As I said, I am not a programmer engineer, so I am not able to go much further as this in what could be happening. The only thing I know is that, if the cache is not what is causing the bug, at least the reason is being stored there from time to time. Maybe is a thing that is happening due to my PC configurations?
I'll put here my PC configuration so maybe someone can see if it has any relation:

-I9 12900HX
-32 GB of RAM
-GeForce RTX 3080 Ti Laptop GPU
-Notebook HP Omen
-Windows 10

Hope it helps to solve it. :/

**Comment by KoBeWi** [comment]
Wait, so you are getting this error in the editor?
Can you reproduce in 4.1 dev?

**Comment by Firezac** [comment]
I apologize for the delay in response. 
Regarding the editor issue, it was occurring before but I haven't been able to replicate it after deleting the project and engine cache. Previously, the project settings, editor settings, or export windows would randomly crash, and attempting to open those windows would cause the engine to crash. However, the bug still exists in the editor, but it no longer blocks or crashes those windows. 
As for the 4.1 dev, the bugs were still present even after deleting the cache. Currently, the bug appears in the Debugger when I open my project, but it doesn't block any functions even after deleting all the caches. I was considering recreating my project in Godot 4.1 Dev to see if starting from scratch would fix the issue. However, I changed my physics_tics_per_second value to accommodate a parkour mechanic, and when I switched to 4.1, my physics became completely broken and I am unsure how to properly regulate it.

**Comment by LuisHdezz** [comment]
I can confirm exactly everything you say about this error, the same thing just happened to me. I changed, deleted and created a new project, this error still appears. I even tried several versions of Godot 4x and the same thing happens in all of them. The physics are broken or do not work correctly. I was trying to create a pixel art platform game in the latest version of Godot: 4.1.1
I haven't tried clearing the cache yet. If anyone found a solution for this, I would appreciate reporting it here.

**Comment by LuisHdezz** [comment]
> I apologize for the delay in response. Regarding the editor issue, it was occurring before but I haven't been able to replicate it after deleting the project and engine cache. Previously, the project settings, editor settings, or export windows would randomly crash, and attempting to open those windows would cause the engine to crash. However, the bug still exists in the editor, but it no longer blocks or crashes those windows. As for the 4.1 dev, the bugs were still present even after deleting the cache. Currently, the bug appears in the Debugger when I open my project, but it doesn't block any functions even after deleting all the caches. I was considering recreating my project in Godot 4.1 Dev to see if starting from scratch would fix the issue. However, I changed my physics_tics_per_second value to accommodate a parkour mechanic, and when I switched to 4.1, my physics became completely broken and I am unsure how to properly regulate it.



Can you confirm if you were able to solve the problem with the solution I found?

**Comment by Firezac** [comment]
So, after all this time, I discovered something that can be interesting for all this situation.
Other similar errors occurred in my project that was given a similar conflict in my project. I discovered that Godot can crash your project if you open your project in one window and then open it again in another without closing the first one. Probably it creates some corruptive files in conflict in cache. 
The easy solution I found:
-Create a new game without anything inside.
-Copy the .godot folder from it
-Delete the .godot folder from your project and paste the one you copy (delete it completely, don't overwrite it)
I don't know if it works for all situations, but for sure will solve the majority of them.

**Comment by dimi3tron** [comment]
This is still an issue in 4.2.
It seems to be an editor issue, with no effects on the actual game.
Running the game straight from command line does not display the error.

I have "Save on focus loss" enabled + I switch between two editors (different projects).

It's a bit unclear when exactly the error is triggered, but focus loss would be my best bet.

**Comment by DriNeo** [comment]
I noticed that problem on Godot 3.5.2 stable

**Comment by LuisHdezz** [comment]
> This is still an issue in 4.2. It seems to be an editor issue, with no effects on the actual game. Running the game straight from command line does not display the error.
> 
> I have "Save on focus loss" enabled + I switch between two open editors (different projects).
> 
> It's a bit unclear when exactly the error is triggered, but focus loss would be my best bet.
> 
> _EDIT_: disabling "Save on focus loss" seems to fix the issue

Exact. It's what worked for me 👌

**Comment by PerMalmberg** [comment]
Still present in 4.2.1.

MRP attached, Just focus/unfocus the Godot window and it is logged.

Error only happens when "interface/editor/save_on_focus_loss" is enabled.

[test.zip](https://github.com/godotengine/godot/files/13800882/test.zip)

**Comment by KoBeWi** [comment]
Seems like `interface/editor/save_on_focus_loss` is bugged in general, in any project. Unfocusing the window will cause error, because the progress dialog is not inside tree and it fails to enter tree, because it tries to do so doing notification. I tried delaying it, but you can't use `call_deferred()` and waiting for a process frame causes stack overflow, because EditorProgressDialog calls `Main::iteration()` 😬

I eventually managed to fix the error, but `save_on_focus_loss` is still unusable. The editor regains focus after saving scene, so you can't switch to another window at all. I think it needs some rework.
Here's a patch that fixes the error:
[fix_focus_error.txt](https://github.com/godotengine/godot/files/13801432/fix_focus_error.txt)

**Comment by peterlamber** [comment]
I noticed this happens only on Windows platforms, I work on the same project on both linux(ubuntu) and a windows setup, windows exhibits this behavior, linux doesn't. 
There are also more file locking errors I sometimes encounter only on windows. 
Might has something to do with a lack of permissions

**Comment by majdisorder** [comment]
> I noticed this happens only on Windows platforms, I work on the same project on both linux(ubuntu) and a windows setup, windows exhibits this behavior, linux doesn't. There are also more file locking errors I sometimes encounter only on windows. Might has something to do with a lack of permissions

It seems like a file locking issue

**Comment by Vennnot** [comment]
Started happening with my Godot as well in new projects with 0 changes in them.

**Comment by theDoktorJ** [comment]
I can also confirm this is happening in 4.2.1 when auto save on focus loss is enabled.  Once it is disabled. the error disappears.  I made numerous changes to my project trying to figure it out, only to find it's an editor bug.  I am on Windows 10.

**Comment by elbroandrew** [comment]
I can confirm this issue happens even when I do not enable 'save on focus'.

**Comment by 2fd5** [comment]
I started getting this error in `v4.3.dev5.official [c9c17d6ca]`
```
Parent node is busy setting up children, `add_child()` failed. Consider using `add_child.call_deferred(child)` instead.
```
I do not have `save on focus` enabled and never had.

This error appears in editor on load and also in release build.

I have not been able to pin point at which stage this error started appearing and to which code change it relates to. I might add more info later.

It would be useful in the mean time if the error message can give more context, as what have caused it.

**Comment by Anixias** [comment]
For me, this has been happening for the last few 4.3 dev versions. It seems to only happen once on project open, but it may have happened once or twice after that.

I'm running `v4.3.dev5.mono.official [89f70e98d]` right now.

```
Parent node is busy setting up children, `add_child()` failed. Consider using `add_child.call_deferred(child)` instead.
scene/main/window.cpp:1718 - Condition "!is_inside_tree()" is true.
```

**Comment by 2fd5** [comment]
I am not getting this error any more on `v4.3.dev6.official [64520fe67]` is it same for others ?

**Comment by akien-mga** [comment]
The issue with "Save on Focus Loss" is a duplicate of #73765 and has just been fixed (for 4.3.beta1, should still be reproducible in 4.3.dev6).

Then it sounds some users had another recent issue without "Save on Focus Loss" in 4.3.dev5, which appears to be fixed in 4.3.dev6.

So this can be closed IMO.

**Comment by Anixias** [comment]
Actually, I'm still getting the error on 4.3.dev6.mono, every time I open a project.

**Comment by AThousandShips** [comment]
> Actually, I'm still getting the error on 4.3.dev6.mono, every time I open a project.

Yes it's been solved after dev6, see https://github.com/godotengine/godot/issues/77095#issuecomment-2104290472
